proposals/nnnn-inferring-senable-for-methods

```diff
@@ -0,0 +1,212 @@
+
+```
+struct User : Sendable {
+  func updatePassword (new: String, old:String) -> Bool { /* update password*/ return true}
+}
+
+**let** unapplied: **@Sendable** (User) → ((String, String) → Bool) = User.updatePassword // no error
+
+**let** partial: **@Sendable** (String, String) → Bool = User().updatePassword // no error
+```
+
+
+
+## Detailed design
+
+This proposal includes five changes to `Sendable` behavior. 
+
+The first two are what we just discussed regarding partial and unapplied function values.
+
+```
+struct User : Sendable {
+  var address
+  var password
+  
+  func changeAddress () {/*do work*/ }
+}
+```
+
+1. The inference of `@Sendable` for unapplied references to methods of a Sendable type. 
+
+```
+**let** unapplied: **@Sendable** (User) → ((String, String) → Void) = User.changeAddress // no error
+```
+
+1. The inference of `@Sendable` for partially-applied methods of a Sendable type.
+
+```
+**let** partial: **@Sendable** (String, String) → Void = User().changeAddress // no error
+```
+
+The next few are: 
+
+1. The inference of `@Sendable`  when referencing non-local functions.
+
+Unlike closures, which retain the captured value, global functions can't capture any variables - because global variables are just referenced by the function without any ownership. With this in mind there is no reason not to make these, Sendable by default.
+
+```
+func doWork() -> Int {
+  `Int.random(in: 1..<42)`
+}
+
+Task<Int, Never>.detached(priority: **nil**, operation: doWork) // Converting non-sendable function value to '@Sendable () async -> Void' may introduce data races
+```
+
+Currently, trying to start a `Task` with the global function `doWork` will cause an error complaining that the function is not `Sendable`. This should compile with no issue.  
+
+1. Prohibition of marking methods `@Sendable` when the type they belong to is not `@Sendable`.
+    1. class C {
+            var random: Int = 0 // random is mutable so `C` can't be checked sendable
+            
+            @Sendable func generateN() async -> Int { //error: adding @Sendable to function of non-Senable type prohibited
+                 random = Int.random(in: 1..<100)
+                 return random 
+            }
+        }
+        
+        Task.detached {
+           let num = C()
+           let n = await num.generateN()
+           num.random = 42 // accessing the `random` var while generateN is mutating it
+        }
+    2. If we move the previous work we wanted to do into a class that stores the random number we generate as a mutable value, we could be introducing a data race by marking the function responsible for this work `@Sendable` . Doing this should be prohibited by the compiler. 
+
+Since `@Sendable` attribute will be automatically determined with this proposal, you don’t have to explicitly write it on function declarations.
+
+## Source compatibility
+
+No impact.
+
+## Effect on ABI stability
+
+This would impact the mangling of function names. 
```

Comment:
Um…how so? Are these impacts going to be a problem in practice?

Reply:
I think this comment belongs in the "Impact on adoption" section and needs elaboration. If you remove an explicit `@Sendable` from a method, the mangling of that method will change. So if someone previously had written `@Sendable` and now wants to remove it to "adopt" this new inference instead, they may need to consider the mangling change.